### ISSUE TITLE
Image backgrounds in GTK3 theme do not actually work

### ISSUE THREAD

**Issue by lukefromdc** [mate-desktop/mate-panel#595]
#### Expected behaviour
An image background for PanelToplevel or .mate-panel-menu-bar set in the GTK theme should be displayed on the panel, same as if it was a user-set custom background. In GTK2 builds of GNOME 2, some themes used an image background, notably UbuntuStudio. 

#### Actual behaviour
An image background will show up fine on the PanelPlugs if they are specified or .mate-panel-menu-bar is used to specify it, but on the panel toplevel it will appear "stretched thin" and almost transparent.  

#### Steps to reproduce the behaviour
Install mate-panel built with GTK 3.22 (maybe 3.18 or later-not tested), and modify the selected GTK theme to set an image backgrond on PanelToplevel and on  #PanelPlug, or with the image background set on on .mate-panel-menu-bar . Use an opaque image to most clearly show the problem.  Ensure the panel background is set to "non-use system theme" and restart mate-panel.

#### MATE general version
1.19.0 from git master as of 6-22-2017

#### Package version
locally compiled

#### Linux Distribution
Debian Unstable with locally built gtk3.22.15 and glib 2.53.1 and Debian's provided libcairo2 version 1.14.8-1

**Comment by lukefromdc** [comment]
There is a codepath in panel-background.c to explicitly handle image backgrounds, but it does not actually workThere is a bug in cairo scaling that makes opaque images translucent if "stretched" too far, rather like rubberized almost opaque material might do in the real world. 

![mate-panel_with_image_bg_in_gtk_theme](https://user-images.githubusercontent.com/14036792/27496608-b19c56c8-5824-11e7-8177-cfe7a1903267.png)

Thus with the current code and a vertical panel, two different rendering problems appear with an image file of a bar with a gradient on it: only the center of the gradient shows on the panel toplevel, and it is nearly transparent due to being stretched by a factor of 1/1920 on my monitor, which apparently means it is pulled to 1920 times its original length-and not all pixels get the color. In GTK2 with circa 2008 cairo what appears to have been this same code worked, and the original UbuntuStudio theme got away with a panel background image just 1 px wide which the code rendered across the entire screen.

 Here is the offending part of "panel_background_prepare" in panel-background.c:

	```
case PANEL_BACK_NONE:
		if (background->default_pattern) {
			/* the theme background-image pattern must be scaled by
			* the width & height of the panel so that when the
			* backing region is cleared
			* (gdk_window_clear_backing_region), the correctly
			* scaled pattern is used */
			cairo_matrix_t m;

			cairo_matrix_init_translate (&m, 0, 0);
			cairo_matrix_scale (&m,
					    1.0 / background->region.width,
					    1.0 / background->region.height);
			cairo_pattern_set_matrix (background->default_pattern, &m);

			gdk_window_set_background_pattern (background->window,
						       background->default_pattern);

```
This part looks quite odd, containing math that scales by 1 divided by the panel dimensions, which with the code in question scales up and not down. My guess is some older version of cairo responded to this by stretching to the final dimensions and then stopping. As of now it computes the true stretch, then renders the portion that fits in the panel-and does not render every pixel if the image is highly stretched.

```
cairo_matrix_scale (&m,
					    1.0 / background->region.width,
					    1.0 / background->region.height);
```

As an experiment, I was able to set a background image in my GTK 3 theme using a single known image file, then hardcode the width and height of that image in the place of each instance of 1.0 above. 

```
cairo_matrix_scale (&m,
					   /* Need to get IMAGE dimensions to replace hardcoded values here!*/
						1600.0/background->region.width,
						23.0/background->region.height);
```

Giving this result:

![mate-panel_with_image_bg_in_gtk_theme_and_hacked_panel](https://user-images.githubusercontent.com/14036792/27496844-bae11510-5825-11e7-98a1-e85d8f456f47.png)

This even worked as I scaled the panel up and down, but of  course wokrs only with a 1600x23 pixel image, and a very short image cannot be stretched to fit without turning translucent. 

If anyone can point me to a way to get the dimensions of the image set in the GTK theme from the cairo_pattern_t into which it is loaded by the code, I would be able to replace those hardcoded dimensions with that code and the function would work for image file backgrounds of reasonable dimensions. A 4096 pixel length of the panel image would guarantee no need to stretch all the way to 4K monitors and would not turn translucent even on a 6x2 array of such monitors, as that seems to start becoming visible at about a 5-1 stretch. That part of the problem is a bug in cairo, not in mate-panel and a fix there would allow 1px long panel background images in the GTK theme.

As for a gradient background image set in GTK3 without a file, that does not work at all, probably coming up with dimensions of zero.
